components/style/stylist.rs

```diff
@@ -98,63 +247,64 @@ impl DocumentCascadeData {
-            let validity = flusher.origin_validity(origin);
-
-            if validity == OriginValidity::Valid {
-                continue;
-            }
-
-            if origin == Origin::UserAgent {
-                self.precomputed_pseudo_element_decls.clear();
-            }
-
-            if validity == OriginValidity::CascadeInvalid {
-                cascade_data.clear_cascade_data()
-            } else {
-                debug_assert_eq!(validity, OriginValidity::FullyInvalid);
-                cascade_data.clear();
-            }
-        }
+        // First do UA sheets.
+        {
+            if flusher.origin_dirty(Origin::UserAgent) {
+                let mut ua_cache = UA_CASCADE_DATA_CACHE.lock().unwrap();
+                let origin_sheets =
+                    flusher.manual_origin_sheets(Origin::UserAgent);
 
-        for (stylesheet, origin, rebuild_kind) in flusher {
-            let guard = guards.for_origin(origin);
-            let origin = stylesheet.origin(guard);
-            self.per_origin
-                .borrow_mut_for_origin(&origin)
-                .add_stylesheet(
+                let ua_cascade_data = ua_cache.lookup(
+                    origin_sheets,
                     device,
                     quirks_mode,
-                    stylesheet,
-                    guard,
-                    rebuild_kind,
-                    &mut self.precomputed_pseudo_element_decls,
+                    guards.ua_or_user
                 )?;
+
+                ua_cache.expire_unused();
+                self.user_agent = ua_cascade_data;
+            }
         }
 
+        // Now do the user sheets.
+        Self::rebuild_origin(
+            device,
+            quirks_mode,
+            &mut flusher,
+            guards,
+            Origin::User,
+            &mut self.user,
+        )?;
+
+        // And now the author sheets.
+        Self::rebuild_origin(
+            device,
+            quirks_mode,
+            &mut flusher,
+            guards,
+            Origin::Author,
+            &mut self.author,
+        )?;
+
         Ok(())
     }
 
     /// Measures heap usage.
     #[cfg(feature = "gecko")]
     pub fn add_size_of_children(&self, ops: &mut MallocSizeOfOps, sizes: &mut ServoStyleSetSizes) {
-        self.per_origin.user_agent.add_size_of_children(ops, sizes);
-        self.per_origin.user.add_size_of_children(ops, sizes);
-        self.per_origin.author.add_size_of_children(ops, sizes);
+        self.user.add_size_of_children(ops, sizes);
+        self.author.add_size_of_children(ops, sizes);
 
-        for elem in self.precomputed_pseudo_element_decls.iter() {
-            if let Some(ref elem) = *elem {
-                sizes.mStylistPrecomputedPseudos += elem.shallow_size_of(ops);
-            }
-        }
+        // FIXME(emilio): UA_CASCADE_DATA_CACHE is shared, we should do whatever
+        // we do for RuleProcessorCache in Gecko.
```

Comment:
FWIW that is to assign all memory for nsCSSRuleProcessors in the RuleProcessorCache to `explicit/layout/rule-processor-cache`, and any non-cached nsCSSRuleProcessors are assigned to the style set.